Bug: ColumnType.DecimalSize panics with nil deref when wrapper has nil SQLColumnType (driver returns partially-populated wrappers under concurrent DROP COLUMN)

## GORM Playground Link

https://github.com/go-gorm/playground/pull/850

## Description

`migrator.ColumnType` is a wrapper struct with a named field `SQLColumnType *sql.ColumnType`. Several wrapper methods (`DecimalSize`, `Length`, `Nullable`, `ScanType`, and the SQLColumnType branches of `Name`/`DatabaseTypeName`) delegate to that field without checking whether it is `nil`. When a driver populates the wrapper from one source (e.g. `information_schema.columns`) but does not attach the `*sql.ColumnType` from a follow-up `SELECT * FROM <table> LIMIT 1` — because the column-name match between the two queries fails — the returned wrapper has `SQLColumnType == nil`. The next caller (typically `MigrateColumn`) panics on the first delegating method call.

A guard like `if columnType == nil` at the call site does NOT protect: the wrapper is a non-nil value type, and the nil pointer is the named field, not the wrapper itself.

### Source location

Wrapper struct: https://github.com/go-gorm/gorm/blob/v1.31.1/migrator/column_type.go#L8-L24

```go
type ColumnType struct {
    SQLColumnType    *sql.ColumnType
    NameValue        sql.NullString
    DataTypeValue    sql.NullString
    ColumnTypeValue  sql.NullString
    ...
}
```

Vulnerable methods (delegate to `ct.SQLColumnType.X()` without a nil check):

```go
// migrator/column_type.go:71-76
func (ct ColumnType) DecimalSize() (precision int64, scale int64, ok bool) {
    if ct.DecimalSizeValue.Valid {
        return ct.DecimalSizeValue.Int64, ct.ScaleValue.Int64, true
    }
    return ct.SQLColumnType.DecimalSize()  // <-- panics if SQLColumnType is nil
}

// Same shape at: Length (L63-68), Nullable (L79-84), ScanType (L92-97),
// and the delegating branches of Name (L27-32) and DatabaseTypeName (L40-45).
```

First observed call site: https://github.com/go-gorm/gorm/blob/v1.31.1/migrator/migrator.go#L528

### How a driver ends up returning a nil-internal wrapper

The postgres driver builds wrappers in two passes inside [`(Migrator).ColumnTypes`](https://github.com/go-gorm/postgres/blob/v1.6.0/migrator.go#L465-L540):

1. **Pass 1** — populate metadata from `information_schema.columns` into one wrapper per column the catalog reports.
2. **Pass 2** — execute `SELECT * FROM <table> LIMIT 1`, call `rows.ColumnTypes()`, then walk both lists and assign each `*sql.ColumnType` into the matching wrapper's `SQLColumnType` field ([driver/migrator.go:528-538](https://github.com/go-gorm/postgres/blob/v1.6.0/migrator.go#L528-L538)):

   ```go
   for _, columnType := range columnTypes {       // wrappers from pass 1
       for _, c := range rawColumnTypes {         // *sql.ColumnType from pass 2
           if c.Name() == columnType.Name() {
               columnType.(*migrator.ColumnType).SQLColumnType = c
               break
           }
       }
   }
   ```

If the `SELECT` itself errors, the function returns the error and no wrappers escape. The wrapper-with-nil-internal case happens when the `SELECT` **succeeds** but `rawColumnTypes` lacks a column the first pass surfaced — exactly the state produced by a concurrent `DROP COLUMN` between the two passes: `information_schema` (first pass) still observes the column under PostgreSQL's MVCC snapshot, but `SELECT * FROM <table>` (second pass) is issued against the post-DROP state. The wrapper for the dropped column survives with `SQLColumnType == nil`. The first method on it that delegates to `ct.SQLColumnType.X()` panics.

(The opposite race — `ADD COLUMN` between passes — is harmless: the new column simply has no wrapper at all.)

### Reproduction

Playground: https://github.com/go-gorm/playground/pull/850 — `TestBug_DecimalSizeNilDerefOnRace` runs 50 race iterations of `AutoMigrate(&Supplier{})` against concurrent `DROP COLUMN`. Hit rate observed: ~30-40 panics per 50 iterations on a single Postgres container with 4 connections.

```go
type Supplier struct {
    ID       uint   `gorm:"primaryKey"`
    Category string `gorm:"type:varchar(50);not null"`
}

for i := 0; i < 50; i++ {
    db.Exec(`DROP TABLE IF EXISTS suppliers`)
    db.Exec(`CREATE TABLE suppliers (id bigserial PRIMARY KEY, category varchar(20) NOT NULL)`)

    var wg sync.WaitGroup
    wg.Add(2)
    go func() {
        defer wg.Done()
        defer func() { _ = recover() }()
        _ = db.AutoMigrate(&Supplier{})  // wants to widen varchar(20) → varchar(50)
    }()
    go func() {
        defer wg.Done()
        defer func() { _ = recover() }()
        db.Exec(`ALTER TABLE suppliers DROP COLUMN IF EXISTS category`)
    }()
    wg.Wait()
}
```

Without the `defer recover()` in the AutoMigrate goroutine, the process crashes with:

```text
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1]
goroutine 68 [running]:
database/sql.(*ColumnType).DecimalSize(...)
        database/sql/sql.go:3248
gorm.io/gorm/migrator.ColumnType.DecimalSize(...)
        gorm.io/gorm@v1.31.1/migrator/column_type.go:75
gorm.io/gorm/migrator.Migrator.MigrateColumn(...)
        gorm.io/gorm@v1.31.1/migrator/migrator.go:528 +0x69e
```

### Production scenarios

- Rolling deploy: two backend instances of the same service start simultaneously; both call `AutoMigrate` against the same DB.
- DBA running `ALTER TABLE DROP COLUMN` for cleanup while a service starts up.
- Concurrent `AutoMigrate` on multiple models that share a column.

User-visible symptom: pod CrashLoops on boot, opaque nil deref, no actionable log line.

### Suggested fix

The defect is in the wrapper, not at any one call site. Make wrapper methods nil-safe by honoring the existing "use cached `*Value` if valid, otherwise delegate" contract even when `SQLColumnType` itself is nil:

```go
// gorm.io/gorm/migrator/column_type.go
func (ct ColumnType) DecimalSize() (precision int64, scale int64, ok bool) {
    if ct.DecimalSizeValue.Valid {
        return ct.DecimalSizeValue.Int64, ct.ScaleValue.Int64, true
    }
+   if ct.SQLColumnType == nil {
+       return 0, 0, false
+   }
    return ct.SQLColumnType.DecimalSize()
}

// Same `if ct.SQLColumnType == nil { return zero, false }` shape applied to:
// Length, Nullable, ScanType, and the SQLColumnType branches of Name and
// DatabaseTypeName.
```

The `ok bool` already exists on `DecimalSize`, `Length`, `Nullable` — a nil internal pointer can safely degrade to `ok=false` and the caller takes the conservative branch (no ALTER, etc.). `ScanType` returns `reflect.Type`; degrading to nil there is consistent with `database/sql` behavior when the column type is unknown.

### Prior reports of the same nil-`SQLColumnType` family

#5262 (closed 2022) described the same wrapper-with-nil-internal symptom from a different call site. A community comment on that issue from `asherascout` diagnosed the exact mechanism:

> "when using select, some columns are missing in rows, so SQLColumnType may be nil"

Closed without a code fix. The companion driver-side panic (different package, same area of code) is filed at https://github.com/go-gorm/postgres/issues/338.

Related but distinct: #6618 (closed) and #5794 (open) cover concurrent-AutoMigrate race conditions in general (CREATE TABLE collisions); not this DecimalSize-specific deref but evidence the concurrent-migrate model is unaddressed.

### Environment

- Go: 1.22.2
- gorm.io/gorm: v1.31.1
- gorm.io/driver/postgres: v1.6.0
- PostgreSQL: 16.4-alpine
- OS: Linux 6.18

--- a/migrator/column_type.go
+++ b/migrator/column_type.go
@@ -28,7 +28,10 @@ func (ct ColumnType) Name() string {
 	if ct.NameValue.Valid {
 		return ct.NameValue.String
 	}
-	return ct.SQLColumnType.Name()
+	if ct.SQLColumnType != nil {
+		return ct.SQLColumnType.Name()
+	}
+	return ""
 }
 
 // DatabaseTypeName returns the database system name of the column type. If an empty
@@ -41,7 +44,10 @@ func (ct ColumnType) DatabaseTypeName() string {
 	if ct.DataTypeValue.Valid {
 		return ct.DataTypeValue.String
 	}
-	return ct.SQLColumnType.DatabaseTypeName()
+	if ct.SQLColumnType != nil {
+		return ct.SQLColumnType.DatabaseTypeName()
+	}
+	return ""
 }
 
 // ColumnType returns the database type of the column. like `varchar(16)`
@@ -64,7 +70,10 @@ func (ct ColumnType) Length() (length int64, ok bool) {
 	if ct.LengthValue.Valid {
 		return ct.LengthValue.Int64, true
 	}
-	return ct.SQLColumnType.Length()
+	if ct.SQLColumnType != nil {
+		return ct.SQLColumnType.Length()
+	}
+	return 0, false
 }
 
 // DecimalSize returns the scale and precision of a decimal type.
@@ -72,7 +81,10 @@ func (ct ColumnType) DecimalSize() (precision int64, scale int64, ok bool) {
 	if ct.DecimalSizeValue.Valid {
 		return ct.DecimalSizeValue.Int64, ct.ScaleValue.Int64, true
 	}
-	return ct.SQLColumnType.DecimalSize()
+	if ct.SQLColumnType != nil {
+		return ct.SQLColumnType.DecimalSize()
+	}
+	return 0, 0, false
 }
 
 // Nullable reports whether the column may be null.
@@ -80,7 +92,10 @@ func (ct ColumnType) Nullable() (nullable bool, ok bool) {
 	if ct.NullableValue.Valid {
 		return ct.NullableValue.Bool, true
 	}
-	return ct.SQLColumnType.Nullable()
+	if ct.SQLColumnType != nil {
+		return ct.SQLColumnType.Nullable()
+	}
+	return false, false
 }
 
 // Unique reports whether the column may be unique.
@@ -93,7 +108,10 @@ func (ct ColumnType) ScanType() reflect.Type {
 	if ct.ScanTypeValue != nil {
 		return ct.ScanTypeValue
 	}
-	return ct.SQLColumnType.ScanType()
+	if ct.SQLColumnType != nil {
+		return ct.SQLColumnType.ScanType()
+	}
+	return nil
 }
 
 // Comment returns the comment of current column.

